### PR TITLE
[tune] Expose an evaluate interface for Trainable

### PR DESCRIPTION
**Pull request by richardliaw**
TODO:
 - [ ] Move PBT code onto using `evaluate`

### CODE REVIEW

File: python/ray/tune/trainable.py [hunk 1]
```diff
@@ -349,9 +359,13 @@ def stop(self):
         self._result_logger.close()
         self._stop()
 
-    def _train(self):
+    def _train(self, evaluate=False):
```
Comment (ericl) [inline python/ray/tune/trainable.py]: Defaulting to true is more backwards compatible.

**Review by ericl — CHANGES_REQUESTED** [review]

File: python/ray/tune/trainable.py (hunk 1 above)
Comment (richardliaw) [reply python/ray/tune/trainable.py]: In this PR, `evaluate` should only be True if you're running evaluation and not training

Comment (richardliaw) [reply python/ray/tune/trainable.py]: Unless you're suggesting `def _train(self, update=True)`

**Comment by ericl** [comment]
Yes, since the existing behaviour is with true as default.

File: python/ray/tune/trainable.py (hunk 1 above)
Comment (richardliaw) [inline python/ray/tune/trainable.py]:
```suggestion
    def _train(self, update=True):
```

File: python/ray/tune/trainable.py [hunk 2]
```diff
@@ -206,6 +206,16 @@ def train(self):
 
         return result
 
+    def evaluate(self):
+        result = {}
+        try:
+            result = self._train(evaluate=True)
```
Comment (richardliaw) [inline python/ray/tune/trainable.py]:
```suggestion
            result = self._train(update=False)
```

File: python/ray/tune/trainable.py [hunk 3]
```diff
@@ -349,9 +359,13 @@ def stop(self):
         self._result_logger.close()
         self._stop()
 
-    def _train(self):
+    def _train(self, update=True):
```
Comment (ericl) [inline python/ray/tune/trainable.py]: This should be evaluate=True right?

File: python/ray/tune/trainable.py [hunk 4]
```diff
@@ -206,6 +206,16 @@ def train(self):
 
         return result
 
+    def evaluate(self):
```
Comment (ericl) [inline python/ray/tune/trainable.py]: Why is this a separate method? Isn't the evaluate flag on train enough?

**Review by ericl — CHANGES_REQUESTED** [review]

File: python/ray/tune/trainable.py (hunk 4 above)
Comment (richardliaw) [reply python/ray/tune/trainable.py]: This was proposed (to which you agreed here) https://groups.google.com/d/msg/ray-dev/S8cN4IKrdWo/liYwLA0-CgAJ

Comment (richardliaw) [reply python/ray/tune/trainable.py]: Models should be able to be evaluated without updating - similar to https://keras.io/models/model/#evaluate

File: python/ray/tune/trainable.py (hunk 3 above)
Comment (richardliaw) [reply python/ray/tune/trainable.py]: OK, what about `validate=True`?

`evaluate` is not the right keyword... in most frameworks (keras, pytorch) `evaluate` is a separate concept than `train`, and `validate` is usually the (sub) concept tied to `train`

**Comment by ericl** [comment]
How about this then.
```
def _evaluate(self):
    """Returns additional evaluation metrics for this model.

    Some scheduling algorithms such as PBT only require
    evaluation metrics every N iterations. Hence, you can
    speed up PBT training by moving expensive model evaluation
    logic from _train() to _evaluate().

    Returns:
        dict with metrics to merge with the train() result.
    """
    return {}

def _train_and_evaluate(self, evaluate=True):
    """(Private method for tune to use)"""

    result = self.train()
    if evaluate:
        result.update(self._evaluate())
    return result
```

**Comment by timonbimon** [comment]
Is this still something that is planned to be added to tune? :)

**Comment by richardliaw** [comment]
Oh, hm I haven't cleaned it up yet.

I promise after June 10 all these stale PRs will be closed :) On the other
hand, feel free to fork off and take ownership of this feature!

**Pull request closed without merging**